### PR TITLE
Make Meter a struct, simplify the global Meter

### PR DESCRIPTION
**Pull request by jmacd**
This PR makes the API more sustainable, making `Meter` a strucrt and requiring that SDKs use the `MeterImpl` interface.  The `global` package simplified, and this will make it easier to add batch observer callbacks and new instruments.

A couple of related simplifications:
- Meter{} is a no-op, no more NoopMeter
- Observer result types become structs, no need for an interface.

### CODE REVIEW

File: api/global/internal/meter.go [hunk 1]
```diff
@@ -264,29 +239,23 @@ func (bound *syncHandle) Unbind() {
 
 // Async delegation
 
-func (m *meter) newAsync(desc metric.Descriptor, constructor func(metric.Meter) (metric.AsyncImpl, error)) (metric.AsyncImpl, error) {
+func (m *meterImpl) NewAsyncInstrument(desc metric.Descriptor,
```
Comment (MrAlias) [inline api/global/internal/meter.go:242]: Is the line-break inclusion intended here?

File: api/global/internal/meter.go [hunk 2]
```diff
@@ -50,21 +50,22 @@ type meterProvider struct {
 	delegate metric.Provider
 
 	lock   sync.Mutex
-	meters map[string]*meter
+	meters map[string]*meterEntry
```
Comment (MrAlias) [inline api/global/internal/meter.go:53]: This should be moved to the `registry` package to unify the location of this type of functionality.

Likely for a separate PR though.

File: api/global/internal/meter.go [hunk 3]
```diff
@@ -50,21 +50,22 @@ type meterProvider struct {
...
+	delegate unsafe.Pointer // (*metric.MeterImpl)
 
 	lock       sync.Mutex
-	registry   map[string]metric.InstrumentImpl
 	syncInsts  []*syncImpl
 	asyncInsts []*asyncImpl
 }
 
+type meterEntry struct {
+	registry metric.MeterImpl
```
Comment (MrAlias) [inline api/global/internal/meter.go:65]: `registry` is an odd name. Would `unique` be okay instead?

File: api/metric/api.go [hunk 4]
```diff
@@ -135,38 +135,99 @@ func (d Descriptor) LibraryName() string {
...
+// duplicate registration).
+func (m Meter) NewFloat64Measure(name string, opts ...Option) (Float64Measure, error) {
+	return wrapFloat64MeasureInstrument(
+		m.newSync(name, MeasureKind, Float64NumberKind, opts))
+}
+
+// RegisterInt64Observer creates a new integer Observer instrument
+// with the given name, customized with options.  May return an error
+// if the name is invalid (e.g., empty) or improperly registered
+// (e.g., duplicate registration).
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/metric/api.go:194]: Done.

File: api/metric/api.go [hunk 5]
```diff
@@ -135,38 +135,99 @@ func (d Descriptor) LibraryName() string {
...
+		m.newAsync(name, ObserverKind, Int64NumberKind, opts,
+			func(observe func(Number, []core.KeyValue)) {
+				callback(Int64ObserverResult{observe})
+			}))
+}
+
+// RegisterFloat64Observer creates a new floating point Observer with
+// the given name, customized with options.  May return an error if
+// the name is invalid (e.g., empty) or improperly registered (e.g.,
+// duplicate registration).
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/metric/api.go]: Done.

File: api/global/internal/meter.go (hunk 1 above)
Comment (jmacd) [reply api/global/internal/meter.go:242]: It wasn't perfect. It was feeling long to me, so It's now written as:
```
func (m *meterImpl) NewAsyncInstrument(
	desc metric.Descriptor,
	callback func(func(metric.Number, []core.KeyValue)),
) (metric.AsyncImpl, error) {
```
I'm happy to go back to one line though if you think we should.

File: api/global/internal/meter.go (hunk 3 above)
Comment (jmacd) [reply api/global/internal/meter.go]: Done.

File: api/global/internal/meter.go (hunk 2 above)
Comment (MrAlias) [reply api/global/internal/meter.go:53]: https://github.com/open-telemetry/opentelemetry-go/issues/714

File: api/metric/sdkhelpers.go [hunk 6]
```diff
@@ -124,131 +107,87 @@ func Configure(opts []Option) Config {
...
-func (m *wrappedMeterImpl) RecordBatch(ctx context.Context, ls []core.KeyValue, ms ...Measurement) {
-	m.impl.RecordBatch(ctx, ls, ms...)
+// MeterImpl returns the underlying MeterImpl of this Meter.
+func (m Meter) MeterImpl() MeterImpl {
+	return m.impl
 }
 
-func (m *wrappedMeterImpl) newSync(name string, metricKind Kind, numberKind NumberKind, opts []Option) (SyncImpl, error) {
+// newSync constructs one new synchronous instrument.
+func (m Meter) newSync(name string, metricKind Kind, numberKind NumberKind, opts []Option) (SyncImpl, error) {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply api/metric/sdkhelpers.go:122]: https://github.com/open-telemetry/opentelemetry-go/issues/716

File: api/global/internal/meter.go (hunk 2 above)
Comment (jmacd) [reply api/global/internal/meter.go:53]: :+1:

File: api/metric/sdkhelpers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply api/metric/sdkhelpers.go:122]: I can't quite see the benefit, or how you would refactor this. Do you mean that, e.g., `MeterImpl.NewSyncInstrument()` should change signature to take `name, metricKind, numberKind, opts` and that all implementations should build the descriptor and check for nil itself?

[the remark this answers is not in the document]
Comment (jmacd) [reply api/metric/sdkhelpers.go:122]: I'm happy to discuss in #716

**Pull request merged**